### PR TITLE
KAFKA-16105: Reset read offsets when seeking to beginning in TBRLMM

### PR DESCRIPTION
**Pull request by AnatolyPopov**
When partition reassignment is happening for a tiered topic in most of the cases it's stuck with RemoteStorageException's on follower nodes saying that it can not construct remote log auxilary state:

```
[2024-01-09 08:34:02,899] ERROR [ReplicaFetcher replicaId=7, leaderId=6, fetcherId=2] Error building remote log auxiliary state for test-24 (kafka.server.ReplicaFetcherThread)
                                         org.apache.kafka.server.log.remote.storage.RemoteStorageException: Couldn't build the state from remote store for partition: test-24, currentLeaderEpoch: 8, leaderLocalLogStartOffset: 209, leaderLogStartOffset: 0, epoch: 0 as the previous remote log segment metadata was not found
                                                 at kafka.server.ReplicaFetcherTierStateMachine.buildRemoteLogAuxState(ReplicaFetcherTierStateMachine.java:259)
                                                 at kafka.server.ReplicaFetcherTierStateMachine.start(ReplicaFetcherTierStateMachine.java:106)
                                                 at kafka.server.AbstractFetcherThread.handleOffsetsMovedToTieredStorage(AbstractFetcherThread.scala:762)
                                                 at kafka.server.AbstractFetcherThread.$anonfun$processFetchRequest$7(AbstractFetcherThread.scala:413)
                                                 at scala.Option.foreach(Option.scala:437)
                                                 at kafka.server.AbstractFetcherThread.$anonfun$processFetchRequest$6(AbstractFetcherThread.scala:332)
                                                 at kafka.server.AbstractFetcherThread.$anonfun$processFetchRequest$6$adapted(AbstractFetcherThread.scala:331)
                                                 at kafka.utils.Implicits$MapExtensionMethods$.$anonfun$forKeyValue$1(Implicits.scala:62)
                                                 at scala.collection.convert.JavaCollectionWrappers$JMapWrapperLike.foreachEntry(JavaCollectionWrappers.scala:407)
                                                 at scala.collection.convert.JavaCollectionWrappers$JMapWrapperLike.foreachEntry$(JavaCollectionWrappers.scala:403)
                                                 at scala.collection.convert.JavaCollectionWrappers$AbstractJMapWrapper.foreachEntry(JavaCollectionWrappers.scala:321)
                                                 at kafka.server.AbstractFetcherThread.processFetchRequest(AbstractFetcherThread.scala:331)
                                                 at kafka.server.AbstractFetcherThread.$anonfun$maybeFetch$3(AbstractFetcherThread.scala:130)
                                                 at kafka.server.AbstractFetcherThread.$anonfun$maybeFetch$3$adapted(AbstractFetcherThread.scala:129)
                                                 at scala.Option.foreach(Option.scala:437)
                                                 at kafka.server.AbstractFetcherThread.maybeFetch(AbstractFetcherThread.scala:129)
                                                 at kafka.server.AbstractFetcherThread.doWork(AbstractFetcherThread.scala:112)
                                                 at kafka.server.ReplicaFetcherThread.doWork(ReplicaFetcherThread.scala:98)
                                                 at org.apache.kafka.server.util.ShutdownableThread.run(ShutdownableThread.java:130)
```
 

Scenario:

A cluster of 3 nodes with a single topic with 30 partitions. All partitions have tiered segments.
Adding 3 more nodes to the cluster and making a reassignment to move all the data to new nodes.
Behavior:
For most of the partitions reassignment is happening smoothly.
For some of the partitions when a new node starts to get assignments it reads __remote_log_metadata topic and tries to initialize the metadata cache on records with COPY_SEGMENT_STARTED. If it's reading such a message for the partition before the partition was assigned to this specific node it ignores the message, so skips the cache initialization and marks the partition as assigned. So reassignment is stuck since COPY_SEGMENT_STARTED is never properly processed.

Expected behavior:
The partitions should not be marked as assigned before the cache is initialized to be able to re-read COPY_SEGMENT_STARTED message and initialize the cache.

Some notes:
This is happening when there are messages in a single metadata partition and the order of the messages does not correspond to the order of assignment. So the follower reads the COPY_SEGMENT_STARTED message, sees that the user partition is not yet assigned to this node, skips the message, and marks the user partition as assigned. Before the metadata record has been read another assignment for another user partition with the same metadata partition but that was already assigned comes in. This leads to one more offset reset to the last read offset, so the metadata records for just assigned partitions are never read since they are in the beginning of the metadata partition. So that skipped COPY_SEGMENT_STARTED seems to be never re-read, so the metadata cache is not initialized.

A proposed solution is to reset the last read offset for metadata partition when the actual offset for that partition is reset to beginning. This allows the reassignment to finish properly.

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/clients/consumer/MockConsumer.java
Comment (gharris1727) [inline clients/src/main/java/org/apache/kafka/clients/consumer/MockConsumer.java]: I don't think we can change this logic, because it would re-deliver the same records over and over, which is a distinct from the real KafkaConsumer that (without seeking) tries to only deliver records once.

I do think that maybe seekToBeginning should clear the records from the reset partitions, to mirror how the real kafka consumer uses FetchState.AWAIT_RESET#hasValidPosition to prevent fetching data from a partition which has just been reset, even if there is data buffered.

Then the test can explicitly re-add the records to re-consume them.

File: storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java
Comment (gharris1727) [inline storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java:247]: Connect's WorkerSinkTask follows a very similar pattern, where consumer.seek is followed by some bookkeeping. We set the internal state to the offset we seek rather than removing it, but i think removing the offset would probably be safer. :+1:

File: storage/src/test/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTaskTest.java
Comment (gharris1727) [inline storage/src/test/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTaskTest.java]: nit: typo tpId0 -> tpId1

File: storage/src/test/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTaskTest.java
Comment (gharris1727) [inline storage/src/test/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTaskTest.java]: this add record is gone now, and that changes the final waitForCondition in this test. Could both of these changes be reverted, or are they integral to the change to the test?

File: storage/src/test/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTaskTest.java
Comment (gharris1727) [inline storage/src/test/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTaskTest.java]: See comment on addRecord.

**Review by gharris1727 — COMMENTED** [review]
Hi @AnatolyPopov thanks for the fix. This is definitely a problem because the internal state of the consumer and the external state has diverged.

I struggled a bit to understand the test scenario and the test changes, but the change itself seems reasonable to me.

File: storage/src/test/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTaskTest.java
Comment (showuon) [reply storage/src/test/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTaskTest.java]: +1. I'd like to see the original test cases are kept, and the tpId3 test is added in the end of the test.

**Review by showuon — COMMENTED** [review]
@AnatolyPopov , thanks for the fix. The change makes sense to me. It's just the change in MockConsumer needs to try not to affect other tests. We hope this fix can be included in v3.8.0, which the deadline will be 2 ~ 3 weeks later. Please let us know if you need any help.

**Comment by AnatolyPopov** [comment]
Thanks @gharris1727 and @showuon for the comments. I'll focus on this starting from tomorrow.

**Comment by showuon** [comment]
@AnatolyPopov , any update for this PR?

**Comment by AnatolyPopov** [comment]
@showuon sorry for the delay, I've got corona last week and still sick. But anyways tried to address the comments, please have a look!

File: storage/src/test/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTaskTest.java
Comment (gharris1727) [inline storage/src/test/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTaskTest.java:275]: I don't like this mock-of-a-mock, but it appears necessary because the ConsumerTaskTest is a multithreaded test.

The existing mocking mechanism for MockConsumer is schedulePollTask, which is ineffective when the background thread is calling poll() _hundreds of thousands of times_. This isn't for you to solve, so I won't block this PR on that change. I've opened https://issues.apache.org/jira/browse/KAFKA-16862 to track the improvement.

**Review by gharris1727 — APPROVED** [review]
LGTM, thanks so much for the fix @AnatolyPopov.

@showuon I'll wait for you to take another pass before merging, since this is eligible to be backported after the 3.8 branch cut.

File: storage/src/test/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTaskTest.java
Comment (AnatolyPopov) [reply storage/src/test/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTaskTest.java:275]: Totally agree about the mock-of-a-mock approach, I don't like it either. But hopefully it's good enough for now and it helps to focus on changes related specifically to this bug in this PR.

**Review by showuon — APPROVED** [review]
LGTM! Thanks for the fix!
@kamalcph , do you want to have another review? I'll merge it tomorrow if no comments from you. Thanks.

**Comment by kamalcph** [comment]
Reviewing the patch, I'll provide my comments by today.

File: storage/src/test/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTaskTest.java
Comment (kamalcph) [inline storage/src/test/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTaskTest.java:304]: nit:

```
Set<TopicIdPartition> partitions = new HashSet<>();
```

**Review by kamalcph — APPROVED** [review]
LGTM, thanks for fixing the issue!

**Comment by kamalcph** [comment]
@gharris1727 @showuon 
Can we merge this patch? We may have to update the commit title.

**Comment by gharris1727** [comment]
Test failures appear unrelated.

**Comment by gharris1727** [comment]
Thank you @AnatolyPopov for the PR, and thanks @showuon and @kamalcph for your reviews!

**Pull request merged**
